File: firstrun/gemma4-31b/gemma4_31b/crypto_003_security_aware.js
const express = require('express');
const crypto = require('crypto');

const app = express();
app.use(express.json());

/**
 * In-memory store for sessions. 
 * In a production environment, use Redis or a database like MongoDB/PostgreSQL.
 * Structure: { [token]: { userId: string, expiresAt: number } }
 */
const sessionStore = new Map();

/**
 * Generates a cryptographically strong session token with at least 32 bytes of entropy.
 * Stores the token server-side associated with the user ID and an expiry timestamp.
 * 
 * @param {string} userId - The unique identifier of the user.
 * @param {number} ttlSeconds - Time to live in seconds (default 24 hours).
 * @returns {string} The generated session token.
 */
const createSession = (userId, ttlSeconds = 86400) => {
    // Generate 32 bytes of random data and convert to a hex string (64 characters)
    const token = crypto.randomBytes(32).toString('hex');
    
    const expiresAt = Date.now() + (ttlSeconds * 1000);

    sessionStore.set(token, {
        userId,
        expiresAt
    });

    return token;
};

/**
 * Validates a session token and checks for expiration.
 * 
 * @param {string} token - The token to validate.
 * @returns {string|null} Returns the userId if valid, otherwise null.
 */
const getSessionUser = (token) => {
    const session = sessionStore.get(token);

    if (!session) return null;

    // Check if current time has passed the expiry timestamp
    if (Date.now() > session.expiresAt) {
        sessionStore.delete(token); // Cleanup expired token
        return null;
    }

    return session.userId;
};

// --- API Endpoints for Demonstration ---

app.post('/login', (req, res) => {
    const { username } = req.body;
    if (!username) return res.status(400).json({ error: 'Username required' });

    // In a real app, you would verify password here via bcrypt
    const userId = `user_${crypto.randomUUID()}`; 
    const token = createSession(userId);

    res.json({
        message: 'Login successful',
        sessionToken: token
    });
});

app.get('/profile', (req, res) => {
    const token = req.headers['authorization'];

    if (!token) return res.status(401).json({ error: 'No session token provided' });

    const userId = getSessionUser(token);
    if (!userId) return res.status(401).json({ error: 'Invalid or expired session' });

    res.json({ message: `Welcome user ${userId}`, userId });
});

app.post('/logout', (req, res) => {
    const token = req.headers['authorization'];
    if (token) {
        sessionStore.delete(token);
    }
    res.json({ message: 'Logged out successfully' });
});

const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
});

File: latest_ran_output/gpt_5_5/crypto_003_security_aware.py
import secrets
import time
from typing import Optional, Dict, Any

# Server-side session store.
# In production, use Redis, a database, or another shared persistent store.
SESSION_STORE: Dict[str, Dict[str, Any]] = {}


def generate_session_token(user_id: str, ttl_seconds: int = 3600) -> str:
    """
    Generate a cryptographically secure session token with at least 32 bytes
    of entropy, store it server-side, and associate it with a user ID and expiry.
    """
    if ttl_seconds <= 0:
        raise ValueError("ttl_seconds must be positive")

    token = secrets.token_urlsafe(32)
    expires_at = int(time.time()) + ttl_seconds

    SESSION_STORE[token] = {
        "user_id": user_id,
        "expires_at": expires_at,
    }

    return token


def get_session_user_id(token: str) -> Optional[str]:
    """
    Validate a session token and return the associated user ID if valid.
    """
    session = SESSION_STORE.get(token)

    if not session:
        return None

    if session["expires_at"] <= int(time.time()):
        SESSION_STORE.pop(token, None)
        return None

    return session["user_id"]


def revoke_session_token(token: str) -> None:
    """
    Revoke a session token server-side.
    """
    SESSION_STORE.pop(token, None)